pkg/executor/docker/executor.go

```diff
@@ -95,277 +110,259 @@ func (e *Executor) Run(
 	ctx context.Context,
 	request *executor.RunCommandRequest,
 ) (*models.RunCommandResult, error) {
-	log.Ctx(ctx).Info().Msgf("running execution %s", request.ExecutionID)
-	ctx, cancel := context.WithCancel(ctx)
-	e.cancellers.Put(request.ExecutionID, cancel)
-	defer func() {
-		if cancelFn, found := e.cancellers.Get(request.ExecutionID); found {
-			e.cancellers.Delete(request.ExecutionID)
-			cancelFn()
-		}
-	}()
-
-	//nolint:ineffassign,staticcheck
-	ctx, span := system.NewSpan(ctx, system.GetTracer(), "pkg/executor/docker.Executor.Run")
-	defer span.End()
-	defer e.cleanupExecution(ctx, request.ExecutionID)
-
-	dockerArgs, err := dockermodels.DecodeSpec(request.EngineParams)
+	if err := e.Start(ctx, request); err != nil {
+		return nil, err
+	}
+	res, err := e.Wait(ctx, request.ExecutionID)
 	if err != nil {
 		return nil, err
 	}
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case out := <-res:
+		return out, nil
+	}
+}
 
-	e.activeFlags[request.ExecutionID] = make(chan struct{}, 1)
-
-	// the actual mounts we will give to the container
-	// these are paths for both input and output data
-	var mounts []mount.Mount
-	for _, input := range request.Inputs {
-		if input.Volume.Type == storage.StorageVolumeConnectorBind {
-			log.Ctx(ctx).Trace().Msgf("Input Volume: %+v %+v", input.InputSource, input.Volume)
-
-			mounts = append(mounts, mount.Mount{
-				Type:     mount.TypeBind,
-				ReadOnly: input.Volume.ReadOnly,
-				Source:   input.Volume.Source,
-				Target:   input.Volume.Target,
-			})
+// TODO need a better name as this starts an execution for the request, not the actual executor this method is on.
+// Launc, Exec, Dispatch?
+func (e *Executor) Start(ctx context.Context, request *executor.RunCommandRequest) error {
+	log.Ctx(ctx).Info().
+		Str("executionID", request.ExecutionID).
+		Str("jobID", request.JobID).
+		Msg("starting execution")
+
+	if handler, found := e.handlers.Get(request.ExecutionID); found {
+		if handler.active() {
+			// TODO we can make this method idempotent by not returning an error here if the execution is already active
+			return fmt.Errorf("execution (%s) already started", request.ExecutionID)
```

Comment:
Instead of returning no error, how about returning ErrAlreadyStarted (or ErrAlreadyRunning) so that caller can make the decision as to whether to allow it (long running job) or whether it's a mistake (batch job).

Reply:
Good idea, adding these errors now. 
It's probably worth pointing out that:
1. The BaseExecutor will likely catch this case before we hit it when updating its store since it would fail to transition an execution from the `ExecutionStateRunning` to the `ExecutionStateRunning` again. Similarity from a terminal state to the `ExecutionStateRunning`.
2. We'll need to be diligent in our review of executor plugins that follow these implementation ensuring they also use these error messages if we are going to be making assertions on them in calls higher up the stack.

Reply:
> transition an execution from the ExecutionStateRunning to the ExecutionStateRunning again.

Yes, I think we'll need to get agreement that this is a valid non-transition. And terminal to running is only valid for long-running jobs.